### PR TITLE
Add string-to-hex and back conversions

### PR DESCRIPTION
**Pull request by FedericoCeratto**
Add parseHexStr and toHex and tests.

### CODE REVIEW

**Comment by data-man** [comment]
@FedericoCeratto 
What about the idea to add support for any type?
Hex-dumping a value of any type would be very useful.

**Comment by FedericoCeratto** [comment]
@data-man yes - if this PR is accepted I'll add more.

**Comment by Yardanico** [comment]
@FedericoCeratto he means creating one proc/macro to print a hex dump of ANY value

**Comment by FedericoCeratto** [comment]
@Yardanico I'll have to add tests and perhaps some logic to dump types in an useful way.

File: lib/pure/strutils.nim [hunk 1]
```diff
@@ -979,6 +992,46 @@ proc parseHexInt*(s: string): int {.noSideEffect, procvar,
...
+  ##    hexToStr("00ff") == "\0\255"
+  ##
+  ## Raises ``ValueError`` for an invalid hex values. The comparison is
+  ## case-insensitive.
+  if s.len mod 2 != 0:
+    raise newException(ValueError, "Incorrect hex string len")
+  result = newString(s.len div 2)
+  var buf = 0
+  for pos, c in s:
+    let val = hex_char_to_value_map[ord(c)].ord
```
Comment (Araq) [inline lib/pure/strutils.nim]: Do we really need this lookup table solution here? The table is full of chr(17).

File: lib/pure/strutils.nim [hunk 2]
```diff
@@ -979,6 +992,46 @@ proc parseHexInt*(s: string): int {.noSideEffect, procvar,
...
+  result = ""
+  for inp in 0..255:
+    let ch = chr(inp)
+    let o =
+      case ch:
+        of '0'..'9': ord('0')
+        of 'a'..'f': ord('a') - 10
+        of 'A'..'F': ord('A') - 10
+        else: -1
+    if o == -1:
```
Comment (Araq) [inline lib/pure/strutils.nim]: if you do not split it into 'case + if' the code becomes simpler.

File: lib/pure/strutils.nim (hunk 1 above)
Comment (FedericoCeratto) [reply lib/pure/strutils.nim]: I ran nimbench, the lookup is faster than alternatives like doing comparisons and subtractions at runtime.

File: lib/pure/strutils.nim (hunk 2 above)
Comment (FedericoCeratto) [reply lib/pure/strutils.nim]: Updated.

File: lib/pure/strutils.nim [hunk 3]
```diff
@@ -1012,6 +1025,43 @@ proc parseHexInt*(s: string): int {.noSideEffect, procvar,
...
+    let ch = chr(inp)
+    let o =
+      case ch:
+        of '0'..'9': inp - ord('0')
+        of 'a'..'f': inp - ord('a') + 10
+        of 'A'..'F': inp - ord('A') + 10
+        else: 17  # indicates an invalid hex char
+    result.add chr(o)
+
+const hex_char_to_value_map = generateHexCharToValueMap()
```
Comment (dom96) [inline lib/pure/strutils.nim]: camelCase please

**Comment by dom96** [comment]
Bonus points if you could fix this as well: https://github.com/nim-lang/Nim/issues/4599

**Comment by dom96** [comment]
Hope @Araq is okay with this.

**Pull request merged**